pkg/rc/replication_controller.go

```diff
@@ -141,97 +148,183 @@ func (rc *replicationController) meetDesires() error {
-				)
-			}
+func (rc *replicationController) ensureConsistency(current map[string]pods.Manifest) error {
+	manifestSHA, err := rc.Manifest.SHA()
+	if err != nil {
+		return err
+	}
+	for node, intentManifest := range current {
+		intentSHA, err := intentManifest.SHA()
+		if err != nil {
+			rc.logger.WithError(err).WithField("node", node).Warn("Could not hash manifest to determine consistency of intent")
+		}
 
-			err := rc.schedule(scheduleOn)
-			if err != nil {
-				return err
-			}
+		if intentSHA == manifestSHA {
+			continue
 		}
-	} else if len(current) > rc.ReplicasDesired {
-		eligible, err := rc.eligibleNodes()
+
+		rc.logger.WithField("node", node).WithField("intentManifestSHA", intentSHA).Info("Found inconsistency in scheduled manifest")
+		rc.schedule(node)
+	}
+
+	return nil
+}
+
+func (rc *replicationController) addPods(current map[string]pods.Manifest) error {
+	currentNodes := nodeNames(current)
+	eligible, err := rc.eligibleNodes()
+	if err != nil {
+		return err
+	}
+
+	// TODO: With Docker or runc we would not be constrained to running only once per node.
+	// So it may be the case that we need to make the Scheduler interface smarter and use it here.
+	possible := sets.NewString(eligible...).Difference(sets.NewString(currentNodes...))
+	toSchedule := rc.ReplicasDesired - len(currentNodes)
+
+	rc.logger.NoFields().Infof("Need to schedule %d nodes out of %s", toSchedule, possible)
+
+	for i := 0; i < toSchedule; i++ {
+		scheduleOn, ok := possible.PopAny()
+		if !ok {
+			return util.Errorf(
+				"Not enough nodes to meet desire: %d replicas desired, %d currentNodes, %d eligible. Scheduled on %d nodes instead.",
+				rc.ReplicasDesired, len(currentNodes), len(eligible), i,
+			)
+		}
+
+		err := rc.schedule(scheduleOn)
 		if err != nil {
 			return err
 		}
+	}
+	return nil
+}
+
+func (rc *replicationController) removePods(current map[string]pods.Manifest) error {
+	currentNodes := nodeNames(current)
+	eligible, err := rc.eligibleNodes()
+	if err != nil {
+		return err
+	}
+
+	// If we need to downsize the number of nodes, prefer any in current that are not eligible anymore.
+	// Additionally, we prefer to unschedule nodes that have manifests that are inconsistent with this RC
+	// Finally, we perform some cleanup on corrupt manifests
+	// TODO: evaluate changes to 'eligible' more frequently
+	preferred := sets.NewString(currentNodes...).Difference(sets.NewString(eligible...))
+
+	rcManifestSHA, err := rc.Manifest.SHA()
+	if err != nil {
+		rc.logger.WithError(err).WithField("rc ID", rc.ID()).Error("This RC has an unhashable manifest!")
+		return err
+	}
+	for node, intentManifest := range current {
+		intentSHA, err := intentManifest.SHA()
+		if err != nil {
+			rc.logger.WithError(err).WithField("node", node).Warn("Could not hash manifest to determine consistency of intent")
+			preferred.Insert(node)
+			continue
+		}
+
+		if intentSHA != rcManifestSHA {
+			preferred.Insert(node)
```

Comment:
are you sure we want to prefer removing a node that has an inconsistent manifest over one that is ineligible? I would think we should unschedule ineligible ones and then let `ensureConsistency` take care of the inconsistencies